### PR TITLE
[CCIP-2049] OffRamp Tests

### PR DESCRIPTION
**Pull request by kalverra**
## Motivation

Rate limiting on the OffRamp solely is more complex and requires significantly more scope of changes to test properly.

## Solution

A separate OffRamp-focused test.

### CODE REVIEW

File: integration-tests/ccip-tests/contracts/contract_models.go [hunk 1]
```diff
@@ -771,6 +771,7 @@ func (pool *TokenPool) SetRemoteChainOnPool(remoteChainSelectors []uint64) error
 	return pool.client.ProcessTransaction(tx)
 }
 
+// SetRemoteChainRateLimits is an offchain function that sets the RateLimiterConfig on the token pool, which controls the Token Pool's rate limit.
```
Comment (b-gopalswami) [inline integration-tests/ccip-tests/contracts/contract_models.go]: Appreciating your effort to add doc comments. 👏

File: integration-tests/ccip-tests/smoke/ccip_test.go [hunk 2]
```diff
@@ -145,31 +149,39 @@ func TestSmokeCCIPRateLimit(t *testing.T) {
 			require.NoError(t, err)
 			tc.lane.Logger.Info().Interface("rate limit", prevRLOnRamp).Msg("Initial OnRamp rate limiter state")
 
-			prevOnRampRLTokenPool, err := src.Common.BridgeTokenPools[0].Instance.GetCurrentOutboundRateLimiterState(nil, tc.lane.Source.DestChainSelector) // TODO RENS maybe?
+			prevOnRampRLTokenPool, err := src.Common.BridgeTokenPools[0].Instance.Latest.PoolInterface.GetCurrentOutboundRateLimiterState(
+				nil, tc.lane.Source.DestinationChainId,
+			) // TODO RENS maybe?
```
Comment (b-gopalswami) [inline integration-tests/ccip-tests/smoke/ccip_test.go:153]: Do we have any recommended formatting settings to keep it consistent across?

File: integration-tests/ccip-tests/smoke/ccip_test.go [hunk 3]
```diff
@@ -387,6 +398,145 @@ func TestSmokeCCIPRateLimit(t *testing.T) {
...
+	t.Parallel()
+
+	log := logging.GetTestLogger(t)
+	TestCfg := testsetups.NewCCIPTestConfig(t, log, testconfig.Smoke)
+
+	// Set the default permissionless exec threshold to 6 minutes so that we can manually execute the transactions faster
+	actions.DefaultPermissionlessExecThreshold = 5 * time.Minute
+	setUpOutput := testsetups.CCIPDefaultTestSetUp(t, log, "smoke-ccip", nil, TestCfg)
+	if len(setUpOutput.Lanes) == 0 {
+		return
```
Comment (b-gopalswami) [inline integration-tests/ccip-tests/smoke/ccip_test.go:417]: I see this is the pattern in most of the test but should we log some warning before we return?

File: integration-tests/ccip-tests/smoke/ccip_test.go [hunk 4]
```diff
@@ -387,6 +398,145 @@ func TestSmokeCCIPRateLimit(t *testing.T) {
...
+	var tests []testDefinition
+	for _, lane := range setUpOutput.Lanes {
+		tests = append(tests, testDefinition{
+			testName: fmt.Sprintf("Network %s to network %s",
+				lane.ForwardLane.SourceNetworkName, lane.ForwardLane.DestNetworkName),
+			lane: lane.ForwardLane,
+		})
+	}
+
+	aggregateRateLimit := big.NewInt(1e16)
```
Comment (b-gopalswami) [inline integration-tests/ccip-tests/smoke/ccip_test.go:432]: Newbie question: Why `1e16` not `1e18`?

File: integration-tests/ccip-tests/smoke/ccip_test.go [hunk 5]
```diff
@@ -387,6 +398,145 @@ func TestSmokeCCIPRateLimit(t *testing.T) {
...
+				Str("Token", limitedSrcToken.ContractAddress.Hex()).
+				Msg("Limited token transfer failed on destination chain (a good thing in this context)")
+
+			// Manually execute the rate limited token transfer and expect a similar error
+			tc.lane.Logger.Info().Str("Wait Time", actions.DefaultPermissionlessExecThreshold.String()).Msg("Waiting for Exec Threshold to Expire")
+			time.Sleep(actions.DefaultPermissionlessExecThreshold) // Give time to exit the window
+			err = tc.lane.ExecuteManually()
+			require.Error(t, err, "There should be errors executing manually at this point")
+			src.UpdateBalance(int64(tc.lane.NumberOfReq), tc.lane.TotalFee, tc.lane.Balance)
+			dest.UpdateBalance(tc.lane.Source.TransferAmount, int64(tc.lane.NumberOfReq), tc.lane.Balance)
```
Comment (b-gopalswami) [inline integration-tests/ccip-tests/smoke/ccip_test.go]: Could you please help me understand why do we need to UpdateBalance at this step?

File: integration-tests/ccip-tests/testconfig/tomls/ccip-default.toml [hunk 6]
```diff
@@ -17,34 +17,37 @@ TTL = '5h'
 # then lanes will be set up between SIMULATED_1 and SIMULATED_2, SIMULATED_1 and SIMULATED_3, SIMULATED_2 and SIMULATED_3
 # default value is ['SIMULATED_1', 'SIMULATED_2'] which means that test will create two private geth networks from scratch and set up lanes between them
 [CCIP.Env.Network]
-selected_networks= ['SIMULATED_1', 'SIMULATED_2']
+selected_networks = ['SIMULATED_1', 'SIMULATED_2']
```
Comment (b-gopalswami) [inline integration-tests/ccip-tests/testconfig/tomls/ccip-default.toml:20]: I guess these changes are auto formatted. If we have settings defined and made it available for possible IDEs, then this will be consistent across.

File: integration-tests/ccip-tests/README.md [hunk 7]
```diff
@@ -2,40 +2,44 @@
 
 Here lives the integration tests for ccip, utilizing our [chainlink-testing-framework](https://github.com/smartcontractkit/chainlink-testing-framework) and [integration-tests](https://github.com/smartcontractkit/ccip/tree/ccip-develop/integration-tests)
 
-## Running the tests
+## Setup the Tests
 
-### Setting up test inputs
+CCIP tests are designed to be highly configurable. Instead of writing many tests to check specific scenarios, the philosophy is to write a few unique tests and make them adjustable through the use of test inputs and configurations. There are a few different ways to set this configuration:
+
+1. Default test input - set via TOML - If no specific input is set; the tests will run with default inputs mentioned in [default.toml](./testconfig/tomls/ccip-default.toml).
```
Comment (b-gopalswami) [inline integration-tests/ccip-tests/README.md:9]: Thanks for fixing the broken link for `default.toml` file. I have noticed that and thought of raising PR on this!

File: integration-tests/ccip-tests/smoke/ccip_test.go (hunk 2 above)
Comment (kalverra) [reply integration-tests/ccip-tests/smoke/ccip_test.go:153]: Go has a [standard style guide](https://google.github.io/styleguide/go/) and it's built in to many IDE auto-formatter tools, we use [golangci-lint](https://github.com/golangci/golangci-lint). I'm not sure why it didn't pick this one up.

File: integration-tests/ccip-tests/smoke/ccip_test.go (hunk 3 above)
Comment (kalverra) [reply integration-tests/ccip-tests/smoke/ccip_test.go:417]: I believe this is a case for running the test in a `remote-test-runner` environment, where we run the test in a K8s pod.

File: integration-tests/ccip-tests/smoke/ccip_test.go (hunk 4 above)
Comment (kalverra) [reply integration-tests/ccip-tests/smoke/ccip_test.go:432]: No real reason other than I wanted it smaller. If we run this on a live network, we will be sending less money around, which means if the test fails to recover the money for whatever reason, it's less of a problem.

File: integration-tests/ccip-tests/smoke/ccip_test.go (hunk 5 above)
Comment (kalverra) [reply integration-tests/ccip-tests/smoke/ccip_test.go]: We probably don't need to in this case. I was just trying to emulate the [previous usage I had seen](https://github.com/smartcontractkit/ccip/blob/ccip-2049/integration-tests/ccip-tests/smoke/ccip_test.go#L643)

File: integration-tests/ccip-tests/testconfig/tomls/ccip-default.toml (hunk 6 above)
Comment (kalverra) [reply integration-tests/ccip-tests/testconfig/tomls/ccip-default.toml:20]: Yes, it's an auto-format, though I'm not sure on the exact one.

File: integration-tests/ccip-tests/Makefile [hunk 8]
```diff
@@ -18,6 +18,8 @@ set_config:
 	@echo "setting secret config with $(secret_toml)"
 	@echo "export BASE64_CCIP_SECRETS_CONFIG=$$(base64 -i $(secret_toml))" >> ./testconfig/override/.env
 	@echo "export TEST_BASE64_CCIP_SECRETS_CONFIG=$$(base64 -i $(secret_toml))" >> ./testconfig/override/.env
+	@echo "BASE64_CCIP_SECRETS_CONFIG=$$(base64 -i $(secret_toml))" > ./testconfig/override/debug.env
+	@echo "TEST_BASE64_CCIP_SECRETS_CONFIG=$$(base64 -i $(secret_toml))" >> ./testconfig/override/debug.env
```
Comment (AnieeG) [inline integration-tests/ccip-tests/Makefile:22]: another` .env `?

File: integration-tests/ccip-tests/actions/ccip_helpers.go [hunk 9]
```diff
@@ -79,10 +78,18 @@ const (
...
 	ChaosGroupNetworkBCCIPGeth        = "CCIPNetworkBGeth"
-	// The higher the load/throughput, the higher value we might need here to guarantee that nonces are not blocked
+)
+
+// TODO: These should be refactored along with the default CCIP test setup to use optional config functions
+var (
+	// DefaultPermissionlessExecThreshold denotes how long the DON will retry a transaction before giving up,
+	// otherwise known as the "Smart Execution Time Window". If a transaction fails to execute within this time window,
+	// the DON will give up and the transaction will need Manual Execution as detailed here: https://docs.chain.link/ccip/concepts/manual-execution#manual-execution
+	// For performance tests: the higher the load/throughput, the higher value we might need here to guarantee that nonces are not blocked
```
Comment (AnieeG) [inline integration-tests/ccip-tests/actions/ccip_helpers.go:88]: is this right assumption?

File: integration-tests/ccip-tests/actions/ccip_helpers.go [hunk 10]
```diff
@@ -79,10 +78,18 @@ const (
...
 	// 1 day should be enough for most of the cases
-	PermissionlessExecThreshold = 60 * 60 * 8 // 8 hr
-	MaxNoOfTokensInMsg          = 50
+	DefaultPermissionlessExecThreshold        = time.Hour * 8
+	DefaultMaxNoOfTokensInMsg          uint16 = 50
+	CurrentVersion                            = "1.5.0-dev"
```
Comment (AnieeG) [inline integration-tests/ccip-tests/actions/ccip_helpers.go]: don't need current version here anymore. We have contract version set for individual contracts now

File: integration-tests/ccip-tests/actions/ccip_helpers.go [hunk 11]
```diff
@@ -2762,6 +2819,9 @@ func (lane *CCIPLane) SendRequests(noOfRequests int, gasLimit *big.Int) error {
 	return nil
 }
 
+// ExecuteManually attempts to execute pending CCIP transactions manually.
+// This is necessary in situations where the transaction is reverted on, or not able to reach,the destination chain,
```
Comment (AnieeG) [inline integration-tests/ccip-tests/actions/ccip_helpers.go]:
```suggestion
// This is necessary in situations where Smart Execution window for that message is over and Offchain plugin will not attempt to execute the message.In such situation any further message from same sender will not be executed until the blocking message is executed by offRamp. So here we call `Execute` on OffRamp manually
```

File: integration-tests/ccip-tests/actions/ccip_helpers.go [hunk 12]
```diff
@@ -3395,8 +3523,7 @@ func (lane *CCIPLane) DeployNewCCIPLane(
 
 	jobParams.P2PV2Bootstrappers = []string{p2pBootstrappersCommit.P2PV2Bootstrapper()}
 
-	// set up ocr2 config
-	err = SetOCR2Configs(commitNodes, execNodes, *lane.Dest)
+	err = SetOCR2Config(commitNodes, execNodes, *lane.Dest)
```
Comment (AnieeG) [inline integration-tests/ccip-tests/actions/ccip_helpers.go:3582]: is removing comment intentional?

File: integration-tests/ccip-tests/contracts/contract_models.go (hunk 1 above)
Comment (AnieeG) [reply integration-tests/ccip-tests/contracts/contract_models.go]: offchain function?

File: integration-tests/ccip-tests/contracts/contract_models.go [hunk 13]
```diff
@@ -1812,7 +1820,141 @@ func (offRamp *OffRamp) UpdateRateLimitTokens(sourceTokens, destTokens []common.
...
+func (offRamp *OffRamp) RemoveRateLimitTokens(ctx context.Context, sourceTokens, destTokens []common.Address) error {
+	if len(sourceTokens) != len(destTokens) {
+		return fmt.Errorf("source and dest tokens must be of the same length")
+	}
+
+	callOpts := &bind.CallOpts{
+		From:    common.HexToAddress(offRamp.client.GetDefaultWallet().Address()),
+		Context: ctx,
+	}
+	existingRateLimitTokens, err := offRamp.Instance.Latest.GetAllRateLimitTokens(callOpts)
```
Comment (AnieeG) [inline integration-tests/ccip-tests/contracts/contract_models.go]: please handle all versions here.

File: integration-tests/ccip-tests/contracts/contract_models.go [hunk 14]
```diff
@@ -1812,7 +1820,141 @@ func (offRamp *OffRamp) UpdateRateLimitTokens(sourceTokens, destTokens []common.
...
+	return offRamp.client.ProcessTransaction(tx)
+}
+
+// RemoveAllRateLimitTokens removes all token pairs from the OffRamp's rate limit.
+func (offRamp *OffRamp) RemoveAllRateLimitTokens(ctx context.Context) error {
+	callOpts := &bind.CallOpts{
+		From:    common.HexToAddress(offRamp.client.GetDefaultWallet().Address()),
+		Context: ctx,
+	}
+	allRateLimitTokens, err := offRamp.Instance.Latest.GetAllRateLimitTokens(callOpts)
```
Comment (AnieeG) [inline integration-tests/ccip-tests/contracts/contract_models.go]: same comment about handling versions.

File: integration-tests/ccip-tests/testconfig/tomls/ccip-default.toml (hunk 6 above)
Comment (AnieeG) [reply integration-tests/ccip-tests/testconfig/tomls/ccip-default.toml:20]: Yeah this probably depends on IDE, in my IDE it formats it to existing version

File: integration-tests/ccip-tests/testsetups/ccip.go [hunk 15]
```diff
@@ -25,27 +25,25 @@ import (
...
 
 	integrationactions "github.com/smartcontractkit/chainlink/integration-tests/actions"
-
-	testutils "github.com/smartcontractkit/chainlink/integration-tests/ccip-tests/utils"
-
 	"github.com/smartcontractkit/chainlink/integration-tests/ccip-tests/actions"
 	"github.com/smartcontractkit/chainlink/integration-tests/ccip-tests/contracts"
 	"github.com/smartcontractkit/chainlink/integration-tests/ccip-tests/contracts/laneconfig"
 	"github.com/smartcontractkit/chainlink/integration-tests/ccip-tests/testconfig"
+	tsconfig "github.com/smartcontractkit/chainlink/integration-tests/ccip-tests/testconfig"
```
Comment (AnieeG) [inline integration-tests/ccip-tests/testsetups/ccip.go]: nit - ccipconfig?

File: integration-tests/ccip-tests/contracts/contract_models.go [hunk 16]
```diff
@@ -1812,7 +1820,141 @@ func (offRamp *OffRamp) UpdateRateLimitTokens(sourceTokens, destTokens []common.
...
+			SourceToken: sourceToken,
+			DestToken:   allRateLimitTokens.DestTokens[i],
+		}
+	}
+
+	opts, err := offRamp.client.TransactionOpts(offRamp.client.GetDefaultWallet())
+	if err != nil {
+		return fmt.Errorf("failed to get transaction opts: %w", err)
+	}
+	tx, err := offRamp.Instance.Latest.UpdateRateLimitTokens(opts, rateLimitTokens, []evm_2_evm_offramp.EVM2EVMOffRampRateLimitToken{})
```
Comment (AnieeG) [inline integration-tests/ccip-tests/contracts/contract_models.go]: same here and applicable all contract methods - Please handle different versions of contracts. If certain functions are not available on certain version please handle that scenario as well. If your test is only applicable on latest, please have that check included in test logic

File: integration-tests/ccip-tests/Makefile (hunk 8 above)
Comment (kalverra) [reply integration-tests/ccip-tests/Makefile:22]: It's helpful for running debug sessions in VSCode.

```json
"envFile": "${workspaceFolder}/ccip-tests/testconfig/override/debug.env"
```

File: integration-tests/ccip-tests/actions/ccip_helpers.go (hunk 9 above)
Comment (kalverra) [reply integration-tests/ccip-tests/actions/ccip_helpers.go:88]: Sorry I'm not sure which part you're referring to?

File: integration-tests/ccip-tests/actions/ccip_helpers.go (hunk 12 above)
Comment (kalverra) [reply integration-tests/ccip-tests/actions/ccip_helpers.go:3582]: Ya, it doesn't really describe anything that the code isn't obviously stating it's doing. We can keep it, but it just seemed redundant to me.

File: integration-tests/ccip-tests/actions/ccip_helpers.go (hunk 9 above)
Comment (AnieeG) [reply integration-tests/ccip-tests/actions/ccip_helpers.go:88]: // For performance tests: the higher the load/throughput, the higher value we might need here to guarantee that nonces are not blocked

File: integration-tests/ccip-tests/contracts/contract_models.go [hunk 17]
```diff
@@ -1664,14 +1667,31 @@ func (onRamp *OnRamp) ApplyPoolUpdates(tokens []common.Address, pools []common.A
...
 type OffRampWrapper struct {
 	Latest *evm_2_evm_offramp.EVM2EVMOffRamp
 	V1_2_0 *evm_2_evm_offramp_1_2_0.EVM2EVMOffRamp
 }
 
-func (offRamp *OffRampWrapper) CurrentRateLimiterState(opts *bind.CallOpts) (RateLimiterConfig, error) {
-	if offRamp.Latest != nil {
-		rlConfig, err := offRamp.Latest.CurrentRateLimiterState(opts)
+// CurrentRateLimiterState retrieves the current rate limiter state for the OffRamp contract
+func (offRamp *OffRamp) CurrentRateLimiterState(opts *bind.CallOpts) (RateLimiterConfig, error) {
```
Comment (AnieeG) [inline integration-tests/ccip-tests/contracts/contract_models.go]: is this change necessary? All other versioned contracts are using the same approach. changing it for just the offRamp will make it an exception.

File: integration-tests/ccip-tests/contracts/contract_models.go [hunk 18]
```diff
@@ -1703,20 +1723,19 @@ type EVM2EVMOffRampExecutionStateChanged struct {
 	Raw            types.Log
 }
 
-type OffRamp struct {
```
Comment (AnieeG) [inline integration-tests/ccip-tests/contracts/contract_models.go:1714]: Same comment as above about introducing this change

File: integration-tests/ccip-tests/actions/ccip_helpers.go (hunk 9 above)
Comment (kalverra) [reply integration-tests/ccip-tests/actions/ccip_helpers.go:88]: That comment was [in the original code](https://github.com/smartcontractkit/ccip/blob/ccip-develop/integration-tests/ccip-tests/actions/ccip_helpers.go#L82). I assume it's still true, but if not we can modify it.

Comment (AnieeG) [reply integration-tests/ccip-tests/actions/ccip_helpers.go:88]: @mateusz-sekara is this still a right assumption?

File: integration-tests/ccip-tests/contracts/contract_models.go [hunk 19]
```diff
@@ -1664,11 +1667,28 @@ func (onRamp *OnRamp) ApplyPoolUpdates(tokens []common.Address, pools []common.A
 	return onRamp.client.ProcessTransaction(tx)
 }
 
+// OffRamp represents the OffRamp CCIP contract on the destination chain
+type OffRamp struct {
```
Comment (AnieeG) [inline integration-tests/ccip-tests/contracts/contract_models.go:1678]: nit: probably a good idea to declare the struct definition right above all its methods if possible.

File: integration-tests/ccip-tests/smoke/ccip_test.go [hunk 20]
```diff
@@ -387,6 +397,141 @@ func TestSmokeCCIPRateLimit(t *testing.T) {
 	}
 }
 
+func TestSmokeCCIPSelfServeRateLimitOffRamp(t *testing.T) {
```
Comment (AnieeG) [inline integration-tests/ccip-tests/smoke/ccip_test.go:400]: Wanna add this to integration-test workflow?

File: integration-tests/ccip-tests/smoke/ccip_test.go [hunk 21]
```diff
@@ -387,6 +397,141 @@ func TestSmokeCCIPRateLimit(t *testing.T) {
...
+	t.Parallel()
+
+	log := logging.GetTestLogger(t)
+	TestCfg := testsetups.NewCCIPTestConfig(t, log, testconfig.Smoke)
```
Comment (AnieeG) [inline integration-tests/ccip-tests/smoke/ccip_test.go:404]: Please check the contract version to run this test. This test will not work with 1.2 ramps

File: integration-tests/ccip-tests/smoke/ccip_test.go (hunk 20 above)
Comment (kalverra) [reply integration-tests/ccip-tests/smoke/ccip_test.go:400]: Yes

**Review by AnieeG — APPROVED** [review]

**Review by jasonmci — APPROVED** [review]
approved

**Review by AnieeG — APPROVED** [review]
🎉
